cleanup(libsinp): Add `fs.*` field class and new `fs.path.*` abstracted fields

mstemm opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines in the [CONTRIBUTING.md](https://github.com/falcosecurity/.github/blob/master/CONTRIBUTING.md) file and learn how to compile Falco from source [here](https://falco.org/docs/source).
2. Please label this pull request according to what type of issue you are addressing.
3. Please add a release note!
4. If the PR is unfinished while opening it specify a wip in the title before the actual title, for example, "wip: my awesome feature"
-->

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

<!--
Please remove the leading whitespace before the `/kind <>` you uncommented.
-->

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area API-version

> /area build

> /area CI

> /area driver-kmod

> /area driver-bpf

> /area driver-modern-bpf

> /area libscap-engine-bpf

> /area libscap-engine-gvisor

> /area libscap-engine-kmod

> /area libscap-engine-modern-bpf

> /area libscap-engine-nodriver

> /area libscap-engine-noop

> /area libscap-engine-source-plugin

> /area libscap-engine-savefile

> /area libscap-engine-udig

> /area libscap

> /area libpman

/area libsinsp

> /area tests

> /area proposals

<!--
Please remove the leading whitespace before the `/area <>` you uncommented.
-->

**Does this PR require a change in the driver versions?**

> /version driver-API-version-major

> /version driver-API-version-minor

> /version driver-API-version-patch

> /version driver-SCHEMA-version-major

> /version driver-SCHEMA-version-minor

> /version driver-SCHEMA-version-patch

<!--
Please remove the leading whitespace before the `/version <>` you uncommented.
-->

**What this PR does / why we need it**:

Add a new set of fields file.path/source/target that work for a variety of different file-related operations. This makes it easier to write general rules that work on files without having to deal with the different arguments that can differ for each rule.

**Which issue(s) this PR fixes**:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
If PR is `kind/failing-tests` please post the related issues/tests in a comment and do not use `Fixes`.
-->

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

<!--
If no, you have to do nothing.
If yes, a release note is required:
Delete `NONE` and enter your extended release note in the block below.
Please note, the release note follows the "conventional commit specification" (https://www.conventionalcommits.org/en/v1.0.0/):
For example: `fix: broken link`.
If the PR requires additional action from users switching to the new release, prepend the string "action required:".
For example, `action required: change the API interface of libscap`.
-->

```release-note
Add a new set of fields file.path/source/target that work for a variety of different file-related operations. This makes it easier to write general rules that work on files without having to deal with the different arguments that can differ for each rule.
```

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/1060#" title="Author self-approved">mstemm</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [mstemm]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

mstemm commented:
Here's an example of how could capture a wide set of file activity in a single rule using these fields:

```- list: file_operation_paths
  items: [/tmp/example-dir]

- macro: file_operation
  condition: (mkdir or rename or remove or open_write or create_symlink or evt.type in (link, linkat))

- rule: Any File Related Operation on Path
  desc: Detect any file operation on a single path
  condition: (file.path pmatch (file_operation_paths) or file.source pmatch (file_operation_paths) or file.target pmatch (file_operation_paths)) and file_operation
  output: Some File Related Operation on Path (evt.type=%evt.type path=%file.path source=%file.source  target=%file.target %user.name=%user.name proc.cmdline=%proc.cmdline proc.pcmdline=%proc.pcmdline container.id=%container.id container.name=%container.name image=%container.image.repository %evt.arg.linkpath %evt.arg.target)
  priority: DEBUG
  source: syscall```

LucaGuerra commented:
Hey Mark! First of all
* I love the tests <3
* This looks like a feature that can make Falco rules easier to use, so I'm interested

One thing that is not too clear to me is that we already have `fd.*` and so having `file.*` could be a bit confusing, and so users should know what one does that the other doesn't and vice versa. So if there is an `fd` should there also always be a `file`? And would that be a source, target or path? Are path and the other two set at the same time? These are simply questions to let me and others better understand the feature.

LucaGuerra commented on userspace/libsinsp/filterchecks.cpp:
```diff
@@ -109,6 +109,256 @@ static inline bool str_match_start(const std::string& val, size_t len, const cha
+sinsp_filter_check_file::sinsp_filter_check_file()
+{
+	m_info.m_name = "file";
+	m_info.m_desc = "Every syscall that has a file descriptor in its arguments has these fields set with information related to the file.";
```
What I was mentioning earlier, this looks the same as `fd.` making the feature harder to understand

mstemm replied:
Good point--I will reword this to it talks about syscalls that act on filesystem paths instead of fds, to emphasize the difference.

lrishi reviewed (commented):
Hi Mark,
this is how things are setup on latest ubuntu:
```
lrishi@lrishi-sdm:/nobackup/lrishi/sysdig$ ls -lha / | grep bin
lrwxrwxrwx   1 root   root      7 Aug  9  2022 bin -> usr/bin
lrwxrwxrwx   1 root   root      8 Aug  9  2022 sbin -> usr/sbin
lrishi@lrishi-sdm:/nobackup/lrishi/sysdig$ ls -lha /usr/bin/python3
lrwxrwxrwx 1 root root 10 Aug 18  2022 /usr/bin/python3 -> python3.10
```

If I install python after starting agent, will `/bin/python3` only resolve to `/bin/python3.10` and not `/usr/bin/python3.10`? Is this acceptable behavior?
OR, 
If cwd of the installing process is `/home/lrishi`, won't this resolve the symlink to /home/lrishi/python3.10?

FedeDP commented:
/milestone 0.12.0

mstemm commented:
Ok, I addressed this round of feedback, and also ensured that in any cases where the path might be relative (symlink events mostly), that the cwd of the threadinfo is prepended to the path.

lrishi commented on userspace/libsinsp/filterchecks.cpp:
```diff
@@ -109,6 +109,265 @@ static inline bool str_match_start(const std::string& val, size_t len, const cha
+	{
+		sinsp_threadinfo* tinfo = evt->get_thread_info();
+
+		m_tstr = tinfo->get_cwd() + m_tstr;
```
```suggestion
                 std::filesystem::path tstr = tinfo->get_cwd() + m_tstr;
		m_tstr = std::filesystem::absolute(tstr).lexically_normal();
```
@mstemm I think we should call lexically normal on the composite path to avoid situations like below from evading rules:
```
root@lrishi-sdm:~# pwd
/root
root@lrishi-sdm:~# ln -s /home/lrishi/td/tf ../home/lrishi/lf
# /root/../home/lrishi/lf
```

mstemm replied:
yeah I can do that. Yay for c++17 and std::filesystem!

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/.github/blob/main/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [56cd529](https://github.com/falcosecurity/libs/commits/56cd529cb22ac2d5c4855c583a3056d2b71bcd04) TEMP: print gcc command lines when building
- [34f8973](https://github.com/falcosecurity/libs/commits/34f8973256b084f2f1566b0cb1848b083d5f2c18) TEMP: Try to get std::filesystem linked into c++

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/.github/blob/main/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [56cd529](https://github.com/falcosecurity/libs/commits/56cd529cb22ac2d5c4855c583a3056d2b71bcd04) TEMP: print gcc command lines when building
- [34f8973](https://github.com/falcosecurity/libs/commits/34f8973256b084f2f1566b0cb1848b083d5f2c18) TEMP: Try to get std::filesystem linked into c++
- [bedd1c2](https://github.com/falcosecurity/libs/commits/bedd1c2c43a1a5bfafdb58ec8bec8c1b2be04dbe) TEMP: try adding -lstdc++fs to linker flags too

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/.github/blob/main/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [56cd529](https://github.com/falcosecurity/libs/commits/56cd529cb22ac2d5c4855c583a3056d2b71bcd04) TEMP: print gcc command lines when building
- [34f8973](https://github.com/falcosecurity/libs/commits/34f8973256b084f2f1566b0cb1848b083d5f2c18) TEMP: Try to get std::filesystem linked into c++
- [bedd1c2](https://github.com/falcosecurity/libs/commits/bedd1c2c43a1a5bfafdb58ec8bec8c1b2be04dbe) TEMP: try adding -lstdc++fs to linker flags too
- [d997350](https://github.com/falcosecurity/libs/commits/d9973503a379f23ffadc63de31cc4cb608589581) TEMP: try more link options

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/.github/blob/main/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [56cd529](https://github.com/falcosecurity/libs/commits/56cd529cb22ac2d5c4855c583a3056d2b71bcd04) TEMP: print gcc command lines when building
- [34f8973](https://github.com/falcosecurity/libs/commits/34f8973256b084f2f1566b0cb1848b083d5f2c18) TEMP: Try to get std::filesystem linked into c++
- [bedd1c2](https://github.com/falcosecurity/libs/commits/bedd1c2c43a1a5bfafdb58ec8bec8c1b2be04dbe) TEMP: try adding -lstdc++fs to linker flags too
- [d997350](https://github.com/falcosecurity/libs/commits/d9973503a379f23ffadc63de31cc4cb608589581) TEMP: try more link options
- [b01a80c](https://github.com/falcosecurity/libs/commits/b01a80c29c3b49e5d65f82ea8dc19fc414a9a690) TEMP: remove STATIC_LINKER_FLAGS

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/.github/blob/main/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [56cd529](https://github.com/falcosecurity/libs/commits/56cd529cb22ac2d5c4855c583a3056d2b71bcd04) TEMP: print gcc command lines when building
- [34f8973](https://github.com/falcosecurity/libs/commits/34f8973256b084f2f1566b0cb1848b083d5f2c18) TEMP: Try to get std::filesystem linked into c++
- [bedd1c2](https://github.com/falcosecurity/libs/commits/bedd1c2c43a1a5bfafdb58ec8bec8c1b2be04dbe) TEMP: try adding -lstdc++fs to linker flags too
- [d997350](https://github.com/falcosecurity/libs/commits/d9973503a379f23ffadc63de31cc4cb608589581) TEMP: try more link options
- [b01a80c](https://github.com/falcosecurity/libs/commits/b01a80c29c3b49e5d65f82ea8dc19fc414a9a690) TEMP: remove STATIC_LINKER_FLAGS
- [da94ffb](https://github.com/falcosecurity/libs/commits/da94ffbe989995b3c45780440a43236a50281d02) TEMP: try adding a different way

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/.github/blob/main/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [65ef178](https://github.com/falcosecurity/libs/commits/65ef17892d4a713bf75c9a34d398b262a0971670) Squash w/ cmake change

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/.github/blob/main/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [65ef178](https://github.com/falcosecurity/libs/commits/65ef17892d4a713bf75c9a34d398b262a0971670) Squash w/ cmake change
- [00676bf](https://github.com/falcosecurity/libs/commits/00676bf0197e5cc70a905a4c8f67b80cabb5c3f6) Squash w/ cmake config

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/.github/blob/main/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [65ef178](https://github.com/falcosecurity/libs/commits/65ef17892d4a713bf75c9a34d398b262a0971670) Squash w/ cmake change
- [00676bf](https://github.com/falcosecurity/libs/commits/00676bf0197e5cc70a905a4c8f67b80cabb5c3f6) Squash w/ cmake config
- [522bb6b](https://github.com/falcosecurity/libs/commits/522bb6bb8ceb657136c750ac716ea1cab49dd0ca) Squash w/ cmake commit

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/.github/blob/main/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [65ef178](https://github.com/falcosecurity/libs/commits/65ef17892d4a713bf75c9a34d398b262a0971670) Squash w/ cmake change
- [00676bf](https://github.com/falcosecurity/libs/commits/00676bf0197e5cc70a905a4c8f67b80cabb5c3f6) Squash w/ cmake config
- [522bb6b](https://github.com/falcosecurity/libs/commits/522bb6bb8ceb657136c750ac716ea1cab49dd0ca) Squash w/ cmake commit

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/.github/blob/main/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [65ef178](https://github.com/falcosecurity/libs/commits/65ef17892d4a713bf75c9a34d398b262a0971670) Squash w/ cmake change
- [00676bf](https://github.com/falcosecurity/libs/commits/00676bf0197e5cc70a905a4c8f67b80cabb5c3f6) Squash w/ cmake config
- [522bb6b](https://github.com/falcosecurity/libs/commits/522bb6bb8ceb657136c750ac716ea1cab49dd0ca) Squash w/ cmake commit
- [6454c6a](https://github.com/falcosecurity/libs/commits/6454c6aec7b0f1497649f2a8bde04c6671189fe0) Squash w/ filtercheck

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/.github/blob/main/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [7c3de51](https://github.com/falcosecurity/libs/commits/7c3de5173f6fe2c680a98d5b37a4b5a3a4ae896e) Squash me

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/.github/blob/main/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [7c3de51](https://github.com/falcosecurity/libs/commits/7c3de5173f6fe2c680a98d5b37a4b5a3a4ae896e) Squash me
- [0432210](https://github.com/falcosecurity/libs/commits/043221033357b0dc5b4b8e421f57c8feefad39ad) Squash me

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

mstemm commented:
@lrishi and @LucaGuerra, if you could look again?

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/1060#pullrequestreview-1411192978" title="Approved">lrishi</a>*, *<a href="https://github.com/falcosecurity/libs/pull/1060#" title="Author self-approved">mstemm</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [mstemm]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

incertum commented:
@mstemm have been thinking about this for several days ... same page re the need to improve the UX. Would you be open to considering my counter proposal?https://github.com/falcosecurity/libs/issues/1134? Thank you!

Please let me know what you all think. @LucaGuerra also mind reading the new proposal and share what your thoughts are?

mstemm commented:
> 

Sure, I commented in the issue.

incertum commented on userspace/libsinsp/filterchecks.cpp:
```diff
@@ -109,6 +110,317 @@ static inline bool str_match_start(const std::string& val, size_t len, const cha
+	std::shared_ptr<gen_event_filter_check> evt_arg_pathname = create_check("evt.rawarg.pathname");
+	std::shared_ptr<gen_event_filter_check> evt_arg_res_eq_0 = create_check("evt.rawarg.res", CO_EQ, "0");
+	std::shared_ptr<gen_event_filter_check> evt_arg_name = create_check("evt.rawarg.name");
+	std::shared_ptr<gen_event_filter_check> evt_fd_name = create_check("fd.name");
```
Preference would be to rather also use the `evt.rawarg*` here to have a clear separation / clean slate between the field classes. Even have the suspicion the old processing for `fd.name` can be cleaned up and simplified in a follow up PR.

mstemm replied:
I could use evt.rawarg for the open variants but not for fchmod/fchown, as they act on a fd but don't have the paths in their own event.

incertum commented on userspace/libsinsp/filterchecks.cpp:
```diff
@@ -109,6 +110,317 @@ static inline bool str_match_start(const std::string& val, size_t len, const cha
+		sinsp_threadinfo* tinfo = evt->get_thread_info();
+
+		std::filesystem::path tstr = tinfo->get_cwd() + m_tstr;
+		m_tstr = std::filesystem::absolute(tstr).lexically_normal().string();
```
This is great! In addition, the old custom implementation for `fd.name` could be cleaned up in a follow up PR to enhance overall robustness and stability. Could we sign you up for that 🙃 ?

incertum commented on userspace/libsinsp/filterchecks.h:
```diff
@@ -231,6 +231,44 @@ friend class chk_compare_helper;
+public:
+	enum check_type
+	{
+		TYPE_NAME = 0,
```
Preference so we can change order more easily in the future if we want.

```suggestion
		TYPE_NAME = 0,
		TYPE_NAMERAW,
		TYPE_SOURCE,
		TYPE_SOURCERAW,
		TYPE_TARGET,
		TYPE_TARGETRAW,
```

mstemm replied:
Looking at other enums in filterchecks.h, around 1/3 define explicit values and 2/3 do not. I agree with jason that explicit values are very slightly better.

But as jason alluded to, the real problem is that the field enums are defined in the header file and the descriptions are in a separate array that has to have the same exact order. I would like to change that someday, but we should change all of them at the same time when we do it.

incertum commented on userspace/libsinsp/filterchecks.cpp:
```diff
@@ -109,6 +110,317 @@ static inline bool str_match_start(const std::string& val, size_t len, const cha
+
+sinsp_filter_check_fspath::sinsp_filter_check_fspath()
+{
+	m_info.m_name = "file";
```
```suggestion
	m_info.m_name = "fspath";
```

mstemm replied:
Yes will fix.

incertum replied:
Mark let's wait until Monday until we get one more round of feedback from @jasondellaluce around final naming throughout as now perhaps `fs` is winning and then maybe change to `fs.path.name` or `fs.path` or `fs.name` we just need to agree.
Just so you don't have to rename stuff 5 times.

mstemm replied:
Updated to `fs.path.*`.

incertum commented on userspace/libsinsp/test/events_file.ut.cpp:
```diff
@@ -571,3 +571,1127 @@ TEST_F(sinsp_with_test_input, signalfd4)
 	ASSERT_FD_GETTERS_NOT_FILE(fdinfo)
 }
+
+TEST_F(sinsp_with_test_input, fspath_field_mkdir)
```
rename this cpp file to `events_fspath.ut.cpp`. Also thanks for the awesome test coverage 🚀 !

jasondellaluce replied:
I agree with Melissa. They are so many that maybe creating a standalone `events_fspath.ut.cpp` would be good for readability. Also, almost all the following test cases all do the same thing basically, just for different event types. Is there a way of having a loop/template/macro setup that would avoid replicating the same test 15-20 times?

mstemm replied:
Sure I will move all the fspath tests to their own file. I also cleaned up the tests to be more dry--all the various arguments are in one location and there are helper methods for the common patterns of enter events/failed exit events not having any fspath values, as well as helpers to more succinctly describe the expected values for each test case.

incertum commented on userspace/libsinsp/test/events_file.ut.cpp:
```diff
@@ -571,3 +571,1127 @@ TEST_F(sinsp_with_test_input, signalfd4)
+	add_default_init_thread();
+	open_inspector();
+	sinsp_evt* evt = NULL;
+	const char *path = "/tmp/test";
```
Can we test with a more complicated test paths like `../\.../../tmp/test` or even some more edge case strategically sprinkled across the tests to more accurately test name and nameraw like also once just the file name without a path etc?

mstemm replied:
Ok added test cases for the complex example `../\.../../tmp/test` as well as a bare name without any path.

incertum commented on userspace/libsinsp/parsers.h:
```diff
@@ -108,6 +108,7 @@ class sinsp_parser
 	void parse_close_exit(sinsp_evt* evt);
 	void parse_thread_exit(sinsp_evt* evt);
 	inline bool detect_and_process_tracer_write(sinsp_evt *evt, int64_t retval, ppm_event_flags eflags);
+	void parse_file_related_exit(sinsp_evt* evt);
```
```suggestion
	void parse_fspath_related_exit(sinsp_evt* evt);
```
?

mstemm replied:
Sure--I wasn't actually thinking of it being specifically related to the fspath filtercheck, but I guess it really is only used for this, so okay.

incertum commented on userspace/libsinsp/filterchecks.cpp:
```diff
@@ -109,6 +110,317 @@ static inline bool str_match_start(const std::string& val, size_t len, const cha
+sinsp_filter_check_fspath::sinsp_filter_check_fspath()
+{
+	m_info.m_name = "file";
+	m_info.m_desc = "Every syscall that has a filesystem path in its arguments has these fields set with information related to the path arguments. This differs from the fd.* fields as it includes syscalls like unlink, rename, etc. that act directly on filesystem paths as compared to opened file descriptors.";
```
Nice!

incertum commented on userspace/libsinsp/CMakeLists.txt:
```diff
@@ -282,6 +282,13 @@ if(NOT WIN32)
+	if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+		if(CMAKE_CXX_COMPILER_VERSION VERSION_LESS 9.0)
+			list(APPEND SINSP_LIBRARIES
+				stdc++fs)
```
❤️

incertum commented:
@mstemm adjusted the PR title, feel free to re-adjust if you don't like it.

jasondellaluce commented on userspace/libsinsp/filterchecks.cpp:
```diff
@@ -109,6 +110,317 @@ static inline bool str_match_start(const std::string& val, size_t len, const cha
+	m_info.m_flags = filter_check_info::FL_WORKS_ON_THREAD_TABLE;
+};
+
+std::shared_ptr<gen_event_filter_check> sinsp_filter_check_fspath::create_check(const char *name,
```
From a design perspective, we're adding yet one more tedious circular dependency. We have the filter factory, which depends on the filterchecks definitions, and then now also one of the filterchecks defintions depending on the filter factory. Wouldn't it be easier at this level to create the filterchecks manually? Most of the ones you use are `evt.*` anyways, so it will be few lines of code for allocating the right filtercheck class and avoid tangling the code dependencies even more than what we already have.

mstemm replied:
Yeah, I can get rid of the factory and just create the `evt.arg` and `fd.*` checks directly.

jasondellaluce commented on userspace/libsinsp/filterchecks.cpp:
```diff
@@ -109,6 +110,317 @@ static inline bool str_match_start(const std::string& val, size_t len, const cha
+
+	m_tstr = "";
+
+	if(m_path_checks.empty())
```
Any compelling reason for doing this lazily instead of at init time?

mstemm replied:
I guess not. What I should really do is make them static so there aren't ~15 additional filterchecks created for every new fs.path.* filtercheck. and now that I'm going to get the factory out of the way, this is possible. So I'll do that.

mstemm replied:
Hrm, I can't make them static as sinsp_filter_check_event and sinsp_filter_check_fd both use m_inspector to check the buffer format. But I will make the set of filterchecks shared, tied to the inspector.

jasondellaluce commented on userspace/libsinsp/filterchecks.cpp:
```diff
@@ -109,6 +110,317 @@ static inline bool str_match_start(const std::string& val, size_t len, const cha
+
+	// First check the success conditions.
+	auto it = m_success_checks.find(evt->get_type());
+	if(it == m_success_checks.end())
+	{
+		return NULL;
+	}
+
+	if(!it->second->compare(evt))
+	{
+		return NULL;
+	}
```
```suggestion
	if(it == m_success_checks.end() || !it->second->compare(evt))
	{
		return NULL;
	}
```

mstemm replied:
Sure, I'll change.

jasondellaluce commented on userspace/libsinsp/filterchecks.cpp:
```diff
@@ -109,6 +110,317 @@ static inline bool str_match_start(const std::string& val, size_t len, const cha
+		return NULL;
+	}
+
+	std::optional<std::string> enter_param;
```
This will create a copy. Can't we have an optional reference wrapper?

mstemm replied:
Yes, replaced with std::optional<std::reference_wrapper<std::string>>.

jasondellaluce commented on userspace/libsinsp/filterchecks.cpp:
```diff
@@ -109,6 +110,317 @@ static inline bool str_match_start(const std::string& val, size_t len, const cha
+		break;
+	}
+
+	if(m_tstr.empty() || m_tstr == "<NA>")
```
Does that mean that we can't have files named `<NA>`? I know this is a long-term issue we have, but do we want to have another feature layer depending on this?

mstemm replied:
I can eliminate this check entirely by changing extract_fspath to have a signature `extract_fspath(sinsp_evt* evt, OUT std::vector<extract_value_t>& values, std::shared_ptr<filtercheck_map_t> checks)` and returning NULL immediately if extract_fspath fails.

jasondellaluce commented on userspace/libsinsp/filterchecks.cpp:
```diff
@@ -109,6 +110,317 @@ static inline bool str_match_start(const std::string& val, size_t len, const cha
+	auto it = checks.find(extract_evt->get_type());
+	if(it == checks.end())
+	{
+		return "";
```
Can't we just return a const char* at this point, and return NULL? This would allow proper null values checking on the invoker, plus avoid the string copy below.

mstemm replied:
See above I changed it to return true/false and fill in a vector of pointers, similar to how `sinsp_filter_check::extract()` (the variant that works on a vector of extracted values) works.

jasondellaluce reviewed (commented):
Left comments. Also, from a design perspective:
- Do we want this field to extract values for other event types, or only for the link, unlink, etc ones? Would you see value in that?
- Shouldn't we add `EF_MODIFIES_STATE` to `PPME_SYSCALL_LINK_E`, `PPME_SYSCALL_LINKAT_E`, `PPME_SYSCALL_MKDIR_E`, `PPME_SYSCALL_RMDIR_E`, `PPME_SYSCALL_UNLINK_E`, `PPME_SYSCALL_UNLINKAT_E`?

mstemm commented:
fchmod/fchown don't currently work as they act on fds but don't look up the fd and associate it with the event. I fixed that in https://github.com/falcosecurity/libs/pull/1154. I expect we'll merge that first, rebase this PR, and then the tests will start passing.

mstemm commented:
> Do we want this field to extract values for other event types, or only for the link, unlink, etc ones? Would you see value in that?

I'm not sure what you mean--what other event types are you referring to? I think this should fill in fspath.* fields for all (or at least most) syscalls that can act on filesystem paths.

mstemm commented:
> Shouldn't we add EF_MODIFIES_STATE to PPME_SYSCALL_LINK_E, PPME_SYSCALL_LINKAT_E, PPME_SYSCALL_MKDIR_E, PPME_SYSCALL_RMDIR_E, PPME_SYSCALL_UNLINK_E, PPME_SYSCALL_UNLINKAT_E?

What state is modified? This just provides a more consistent way to access the pathname used by all those syscalls.

incertum commented:
@mstemm let's go with below naming, see https://github.com/falcosecurity/libs/issues/1134#issuecomment-1597460761, plz all previous comments re renaming should now reflect our new decision, thank you!

```
fs.path.name
fs.path.nameraw
fs.path.source
fs.path.sourceraw
fs.path.target
fs.path.targetraw
```

In addition, @jasondellaluce I also think changing or adjusting the state table should be done in a different PR as it is unrelated to this one.

mstemm commented on userspace/libsinsp/event.cpp:
```diff
@@ -2850,3 +2850,52 @@ bool sinsp_evt::clone_event(sinsp_evt &dest, const sinsp_evt &src)
+
+void sinsp_evt::save_enter_event_params(sinsp_evt* enter_evt)
+{
+	std::vector<std::string> pnames;
```
Sure, made the per-event type vectors static std::vector<const char*> with a vector pointer that's set based on the event type.

mstemm commented on userspace/libsinsp/event.h:
```diff
@@ -599,6 +607,9 @@ VISIBILITY_PRIVATE
 	std::shared_ptr<sinsp_fdinfo_t> m_fdinfo_ref;
+	// For some exit events, the "path" argument from the
+	// corresponding enter event is stored here.
+	std::map<std::string, std::string> m_enter_path_param;
```
Yes, changed to unordered_map.

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/.github/blob/main/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [022e522](https://github.com/falcosecurity/libs/commits/022e5223a53039f0eb81f4273142bee945370258) Squash w Add a general way to save specific params from enter event
- [d3a4062](https://github.com/falcosecurity/libs/commits/d3a406228b270bbfd864ae6cedc5998569cde0cc) Squash w Add a general way to save specific params from enter event

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

mstemm commented:
Here is the corresponding docs PR: https://github.com/falcosecurity/falco-website/pull/1010.

mstemm commented:
I believe I addressed the latest round of feedback. Also, I rebased to pick up the fchmod/fchown changes that were in https://github.com/falcosecurity/libs/pull/1154. So please take a look again!

jasondellaluce reviewed (approved):
/approve

/milestone 0.12.0

poiana commented:
LGTM label has been added.  <details>Git tree hash: ad0a7f67dda322ec5275e7e788ef112b9a0b6578</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/1060#pullrequestreview-1492763212" title="Approved">jasondellaluce</a>*, *<a href="https://github.com/falcosecurity/libs/pull/1060#pullrequestreview-1411192978" title="Approved">lrishi</a>*, *<a href="https://github.com/falcosecurity/libs/pull/1060#" title="Author self-approved">mstemm</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [jasondellaluce,mstemm]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

incertum reviewed (approved):
/approve

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/1060#pullrequestreview-1493494358" title="Approved">incertum</a>*, *<a href="https://github.com/falcosecurity/libs/pull/1060#pullrequestreview-1492763212" title="Approved">jasondellaluce</a>*, *<a href="https://github.com/falcosecurity/libs/pull/1060#pullrequestreview-1411192978" title="Approved">lrishi</a>*, *<a href="https://github.com/falcosecurity/libs/pull/1060#" title="Author self-approved">mstemm</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [incertum,jasondellaluce,mstemm]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->
